collect_app/src/main/java/org/odk/collect/android/instancemanagement/send/InstanceUploaderActivity.java

```diff
@@ -386,4 +387,14 @@ public void updatedCredentials() {
     public void cancelledUpdatingCredentials() {
         finish();
     }
+
+    private String getReferrerHost() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1) {
+            Uri referrerUri = getReferrer();
+            if (referrerUri != null) {
+                return referrerUri.getHost();
```

Comment:
Do we definitely just want the host here?

Reply:
Looked a bit deeper into this and it looks like the referrer extra needs to be explicitly set: https://developer.android.com/reference/android/content/Intent#EXTRA_REFERRER It sounds like it's set automatically when the referrer is a web app but not when it's an Android app. We may get all nulls but I still think we should try it.

Can we just return `referrerUri.toString`?

Reply:
Oh good:
> If the launching Intent contains an [Intent.EXTRA_REFERRER](https://developer.android.com/reference/android/content/Intent#EXTRA_REFERRER), that will be returned as-is; otherwise, if known, an [android-app:](https://developer.android.com/reference/android/content/Intent#URI_ANDROID_APP_SCHEME) referrer URI containing the package name that started the Intent will be returned. This may return null if no referrer can be identified -- it is neither explicitly specified, nor is it known which application package was involved.

-- https://developer.android.com/reference/android/app/Activity#getReferrer()

I thought I was losing my mind for a minute there, I could have sworn I saw more certainty that it would be set.

Still, `referrerUri.toString` seems like the best thing to return.

Reply:
Ok I changed it to referrerUri.toString if you prefer it.

> It sounds like it's set automatically when the referrer is a web app but not when it's an Android app. We may get all nulls but I still think we should try it.

I tested it with [Collect Intents Tester ](https://github.com/grzesiek2010/collectTester)and it didn't return nulls.